### PR TITLE
wallet: Cleanup and separate BerkeleyDatabase and BerkeleyBatch

### PR DESCRIPTION
**Pull request by achow101**
`BerkeleyBatch` and `BerkeleyDatabase` are kind of messy. The goal of this is to clean up them up so that they are logically separated.

`BerkeleyBatch` currently handles the creation of the `BerkeleyDatabase`'s `Db` handle. This is instead moved into `BerkeleyDatabase` and is called by `BerkeleyBatch`.

Instead of having `BerkeleyEnvironment` track each database's usage, have `BerkeleyDatabase` track this usage itself with the `m_refcount` variable that is present in `WalletDatabase`.

Lastly, instead of having each `BerkeleyEnvironment` store the fileids of the databases open in it, have a global `g_fileids` to track those fileids. We were already checking fileid uniqueness globally (by checking the fileids in every environment when opening a database) so it's cleaner to do this with a global variable.

All of these changes allow us to make `BerkeleyBatch` and `BerkeleyDatabase` no longer be friend classes.

The diff of this PR is currently the same as in ##18971

Requires #19334

### CODE REVIEW

**Review by ryanofsky — APPROVED** [review]
Code review ACK d416ae560e46a4846a3fd5990b7d390d2ef30ec8. Only changes since last review were rebasing after base PR #19334 merge, and adding cs_db lock in BerkeleyDatabase destructor, which should avoid races accessing env->m_databases and env->m_fileids

**Review by ryanofsky — COMMENTED** [review]
Conditional code review ACK e73d0a1f9e28f303024d6ddec2b4e7e980783c9c if some problems below are fixed. Just reviewed last 4 commits here

- [X] 61da2fa6a159b0d318978cb8713f83705bb97035 walletdb: track database file use as m_refcount within BerkeleyDatabase (5/8)
- [X] 33554edc24a51685cfc3ca0053372c253d1fb4b9 walletdb: Move Db->open to BerkeleyDatabase::Open (6/8)
- [X] 8a98a83232fab6ff653c9475bb6e01249eadfa5c walletdb: Use a global g_fileids instead of m_fileids for each env (7/8)
- [X] e73d0a1f9e28f303024d6ddec2b4e7e980783c9c walletdb: Remove BerkeleyBatch friend class from BerkeleyDatabase (8/8)

File: src/wallet/bdb.cpp [hunk 1]
```diff
@@ -245,9 +244,6 @@ BerkeleyEnvironment::BerkeleyEnvironment()
 
 bool BerkeleyEnvironment::Verify(const std::string& strFile)
 {
-    LOCK(cs_db);
-    assert(mapFileUseCount.count(strFile) == 0);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/bdb.cpp:249]: I've added a commit to merge `BerkeleyEnvironment::Verify` and `BerkeleyDatabase::Verify`

File: src/wallet/bdb.cpp [hunk 2]
```diff
@@ -591,10 +588,10 @@ void BerkeleyEnvironment::Flush(bool fShutdown)
         return;
     {
         LOCK(cs_db);
-        std::map<std::string, int>::iterator mi = mapFileUseCount.begin();
-        while (mi != mapFileUseCount.end()) {
-            std::string strFile = (*mi).first;
-            int nRefCount = (*mi).second;
+        bool no_dbs_accessed = true;
+        for (auto& db_it : m_databases) {
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/bdb.cpp:580]: I've changed `m_refcount` to track when the database has been flushed by setting it to `-1`. This is being done everywhere `mapFileUseCount.erase()` was happening.

File: src/wallet/bdb.cpp [hunk 3]
```diff
@@ -633,21 +630,19 @@ bool BerkeleyDatabase::PeriodicFlush()
...
-    for (const auto& use_count : env->mapFileUseCount) {
-        if (use_count.second > 0) return false;
+    for (auto& it : env->m_databases) {
+        if (it.second.get().m_refcount > 0) return false;
     }
 
     // Don't flush if there haven't been any batch writes for this database.
-    auto it = env->mapFileUseCount.find(strFile);
-    if (it == env->mapFileUseCount.end()) return false;
+    if (m_refcount > 0) return false;
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/bdb.cpp]: Changed to use `m_refcount < 0` as mentioned in a previous comment.

File: src/wallet/bdb.cpp [hunk 4]
```diff
@@ -334,13 +334,27 @@ BerkeleyDatabase::~BerkeleyDatabase()
 
 BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bool fFlushOnCloseIn) : pdb(nullptr), activeTxn(nullptr), m_cursor(nullptr), m_database(database)
 {
+    database.AddRef();
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/bdb.cpp:334]: I've moved `m_database.RemoveRef()` to the destructor.

File: src/wallet/bdb.cpp [hunk 5]
```diff
@@ -389,24 +412,10 @@ void BerkeleyDatabase::Open(const char* pszMode)
...
-            // environment when the file is referenced through equivalent but
-            // not obviously identical symlinked or hard linked or bind mounted
-            // paths. In the future a more relaxed check for equal inode and
-            // device ids could be done instead, which would allow opening
-            // different backup copies of a wallet at the same time. Maybe even
-            // more ideally, an exclusive lock for accessing the database could
-            // be implemented, so no equality checks are needed at all. (Newer
-            // versions of BDB have an set_lk_exclusive method for this
-            // purpose, but the older version we use does not.)
-            for (const auto& env : g_dbenvs) {
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/bdb.cpp:388]: Replaced that commit to do as you suggest.

File: src/wallet/bdb.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/wallet/bdb.cpp:580]: re: https://github.com/bitcoin/bitcoin/pull/19335#discussion_r455990656

In commit "walletdb: track database file use as m_refcount within BerkeleyDatabase" (516084a2738364f3922c08d8703fd737541bb2da)

> I've changed `m_refcount` to track when the database has been flushed by setting it to `-1`. This is being done everywhere `mapFileUseCount.erase()` was happening. This allows us to restore the original behavior by checking for `m_refcount < 0`.

Note: I think there is still small change in behavior because m_refcount is initialized to 0 not -1, so never used databases are now treated like used and not yet flushed databases. I don't think the difference actually matters, though.

**Review by ryanofsky — APPROVED** [review]
Code review ACK ccfd611ed1bb45874a2558390de5597dd32d2772 last 5 commits

- [X] 747ca18d5b5f6dd3780de0f543b8a06d52c74169 Combine BerkeleyEnvironment::Verify into BerkeleyDatabase::Verify (5/9)
- [X] 516084a2738364f3922c08d8703fd737541bb2da walletdb: track database file use as m_refcount within BerkeleyDatabase (6/9)
- [X] 93b5f0442bada706aba561dfb5bf3da78101522b walletdb: Move Db->open to BerkeleyDatabase::Open (7/9)
- [X] 0a01ba79a4f7076593b6f65cb3e227455a487219 No need to check for duplicate fileids in all dbenvs (8/9)
- [X] ccfd611ed1bb45874a2558390de5597dd32d2772 walletdb: Remove BerkeleyBatch friend class from BerkeleyDatabase (9/9)

Only changes since previous review are followups from https://github.com/bitcoin/bitcoin/pull/19335#pullrequestreview-449402796. This is a nice cleanup and everything here is an improvement. I do think Batch/Database refcounting interaction is too complicated though and could be made simpler later (https://github.com/bitcoin/bitcoin/pull/19335#discussion_r455438675).

**Review by ryanofsky — APPROVED** [review]
Code review ACK 74507ce71eb61105fb3ae8460999099234ca7b8b. No changes since last review other than rebase

File: src/wallet/bdb.cpp [hunk 6]
```diff
@@ -285,8 +284,7 @@ bool BerkeleyDatabase::Verify(bilingual_str& errorStr)
 
     if (fs::exists(file_path))
     {
-        LOCK(cs_db);
```
Comment (promag) [inline src/wallet/bdb.cpp]: Why was this removed?

Comment (achow101) [reply src/wallet/bdb.cpp]: It was there to guard mapFileUseCount which we've now removed.

**Comment by laanwj** [comment]
Code review ACK 74507ce71eb61105fb3ae8460999099234ca7b8b
